src/thread.cr

```diff
@@ -48,51 +52,49 @@ class Thread
     end
   end
 
-  # All threads, so the GC can see them (GC doesn't scan thread locals)
-  # and we can find the current thread on platforms that don't support
-  # thread local storage (eg: OpenBSD)
-  @@threads = Set(Thread).new
+  {% if flag?(:android) || flag?(:openbsd) %}
+    # no thread local storage (TLS) for OpenBSD or Android,
+    # we use pthread's specific storage (TSS) instead:
+    @@current_key = uninitialized LibC::PthreadKeyT
 
-  {% if flag?(:openbsd) %}
-    @@main = new
+    ret = LibC.pthread_key_create(pointerof(@@current_key), nil)
+    raise Errno.new("pthread_key_create") unless ret == 0
 
     def self.current : Thread
-      if LibC.pthread_main_np == 1
-        return @@main
-      end
-
-      current_thread_id = LibC.pthread_self
-
-      current_thread = @@threads.find do |thread|
-        LibC.pthread_equal(thread.id, current_thread_id) != 0
+      if ptr = LibC.pthread_getspecific(@@current_key)
+        ptr.as(Thread)
+      else
+        raise "BUG: Thread.current returned NULL"
       end
-
-      raise "Error: failed to find current thread" unless current_thread
-      current_thread
     end
 
-    protected def id
-      @th.not_nil!
+    protected def self.current=(thread : Thread) : Thread
+      ret = LibC.pthread_setspecific(@@current_key, thread.as(Void*))
+      raise Errno.new("pthread_setspecific") unless ret == 0
+      thread
     end
+
+    self.current = new
   {% else %}
     @[ThreadLocal]
     @@current = new
 
-    def self.current
+    def self.current : Thread
       @@current
     end
+
+    protected def self.current=(@@current : Thread) : Thread
+    end
   {% end %}
 
   protected def start
-    {% unless flag?(:openbsd) %}
-    @@current = self
-    {% end %}
+    Thread.current = self
```

Comment:
For the main thread the `#start` is never called, so setting the current thread by this line wont happen.
In my WIP for mt-gc at https://github.com/bcardiff/crystal/commit/aedde16f9d012a8d8649b29974e9eec01cb29870 I changed how main fiber / thread will initialize.

On platforms with `ThreadLocal` there will be 2 instances of Thread objects with the current code since the `@@current = new` will be evaluated again.

For now, performing `Thread.current = self` in both constructors seem enough.

Reply:
I now always call `self.current = new` in `Thread`, but maybe it would be better placed in `src/kernel.cr` (eventually).